src/librustc/ty/sty.rs

```diff
@@ -276,6 +279,50 @@ impl<'a, 'gcx, 'acx, 'tcx> ClosureSubsts<'tcx> {
     }
 }
 
+impl<'a, 'gcx, 'tcx> ClosureSubsts<'tcx> {
+    pub fn state_tys(self, def_id: DefId, tcx: TyCtxt<'a, 'gcx, 'tcx>) ->
+        impl Iterator<Item=Ty<'tcx>> + 'tcx
+    {
+        let state = tcx.generator_layout(def_id).fields.iter();
+        let state: Vec<_> = state.map(|d| d.ty.subst(tcx, self.substs)).collect();
```

Comment:
This is the error I get if I just return `state.map(|d| d.ty.subst(tcx, self.substs))`:
```
error[E0477]: the type `core::iter::Map<core::slice::Iter<'_, mir::LocalDecl<'_>>, [closure@src\librustc\ty\sty.rs:290:19: 290:51 tcx:&ty::context::TyCtxt<'a, 'gcx, 'tcx>, self:&ty::sty::ClosureSubsts<'tcx>]>` does not fulfill the required lifetime
   --> src\librustc\ty\sty.rs:287:9
    |
287 |         impl Iterator<Item=Ty<'tcx>> + 'tcx
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: type must outlive the lifetime 'tcx as defined on the impl at 282:1
   --> src\librustc\ty\sty.rs:282:1
    |
282 | / impl<'a, 'gcx, 'tcx> ClosureSubsts<'tcx> {
283 | |     /// This returns the types of the MIR locals which had to be stored across suspension points.
284 | |     /// It is calculated in rustc_mir::transform::generator::StateTransform.
285 | |     /// All the types here must be in the tuple in GeneratorInterior.
...   |
302 | |     }
303 | | }
    | |_^

error[E0495]: cannot infer an appropriate lifetime for capture of `tcx` by closure due to conflicting requirements
   --> src\librustc\ty\sty.rs:290:19
    |
290 |         state.map(|d| d.ty.subst(tcx, self.substs))
    |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the impl at 282:1...
   --> src\librustc\ty\sty.rs:282:1
    |
282 | / impl<'a, 'gcx, 'tcx> ClosureSubsts<'tcx> {
283 | |     /// This returns the types of the MIR locals which had to be stored across suspension points.
284 | |     /// It is calculated in rustc_mir::transform::generator::StateTransform.
285 | |     /// All the types here must be in the tuple in GeneratorInterior.
...   |
302 | |     }
303 | | }
    | |_^
note: ...so that the reference type `&ty::context::TyCtxt<'a, 'gcx, 'tcx>` does not outlive the data it points at
   --> src\librustc\ty\sty.rs:290:9
    |
290 |         state.map(|d| d.ty.subst(tcx, self.substs))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: but, the lifetime must be valid for the lifetime 'tcx as defined on the impl at 282:1...
   --> src\librustc\ty\sty.rs:282:1
    |
282 | / impl<'a, 'gcx, 'tcx> ClosureSubsts<'tcx> {
283 | |     /// This returns the types of the MIR locals which had to be stored across suspension points.
284 | |     /// It is calculated in rustc_mir::transform::generator::StateTransform.
285 | |     /// All the types here must be in the tuple in GeneratorInterior.
...   |
302 | |     }
303 | | }
    | |_^
note: ...so that the type `core::iter::Map<core::slice::Iter<'_, mir::LocalDecl<'_>>, [closure@src\librustc\ty\sty.rs:290:19: 290:51 tcx:&ty::context::TyCtxt<'a, 'gcx, 'tcx>, self:&ty::sty::ClosureSubsts<'tcx>]>` will meet its required lifetime bounds
   --> src\librustc\ty\sty.rs:287:9
    |
287 |         impl Iterator<Item=Ty<'tcx>> + 'tcx
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error(s)
```